File: C++/94.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 //递归

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;
        if(root!=nullptr)
            travel(root,output);
        return output;
    }

    void travel(TreeNode* root, vector<int>& output){
        if(root->left!=nullptr)
            travel(root->left,output);
        output.push_back(root->val);
        if(root->right!=nullptr)
            travel(root->right,output);
    }
};

//迭代
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;
        stack<TreeNode*> temp;
        //if(root==nullptr) return output;
        while(true){
            if(root!=nullptr){
                temp.push(root);
                root=root->left;
            }else{
                if(temp.empty()){
                    break;
                }else{
                    root=temp.top();
                    temp.pop();
                    output.push_back(root->val);
                    root=root->right;
                }
            }
        }
        return output;
    }
};

File: GO/94.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func travel (root * TreeNode, output * []int){
	if(root.Left!=nil){
		travel(root.Left,output)
	}
	*output=append(*output,root.Val)
	if(root.Right!=nil){
		travel(root.Right,output)
	}
}
// solution 1
func inorderTraversal(root *TreeNode) []int {
	var output []int;
	if(root!=nil){
		travel(root,&output)
	}
	return output
}

//solution 2
func inorderTraversal(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    result := inorderTraversal(root.Left)
    result = append(result, root.Val)
    return append(result, inorderTraversal(root.Right)...)
}

//GO 有指针操作 & *

// ...将切片打散传入